ui/hooks/snaps/useTransactionInsightSnaps.js

```diff
@@ -0,0 +1,113 @@
+import { useEffect, useState } from 'react';
+import { useSelector } from 'react-redux';
+import { getTransactionOriginCaveat } from '@metamask/snaps-controllers';
+import { handleSnapRequest } from '../../store/actions';
+import { getPermissionSubjects } from '../../selectors';
+
+const INSIGHT_PERMISSION = 'endowment:transaction-insight';
+
+export function useTransactionInsightSnaps({
+  transaction,
+  chainId,
+  origin,
+  insightSnaps,
+  ///: BEGIN:ONLY_INCLUDE_IN(build-main,build-mmi,build-beta)
+  insightSnapId = '',
+  ///: END:ONLY_INCLUDE_IN
+}) {
+  const subjects = useSelector(getPermissionSubjects);
+
+  const [loading, setLoading] = useState(true);
+  const [data, setData] = useState(undefined);
+  ///: BEGIN:ONLY_INCLUDE_IN(build-flask)
+  const [hasFetchedV2Insight, setHasFetchedV2Insight] = useState(false);
+  ///: END:ONLY_INCLUDE_IN
+  useEffect(() => {
+    let cancelled = false;
+
+    async function fetchInsight() {
+      ///: BEGIN:ONLY_INCLUDE_IN(build-flask)
+      if (hasFetchedV2Insight) {
```

Comment:
This doesn't look like it is enough. Snaps are still being called more than once. You can test by simply logging in the snap onTransaction handler.

Reply:
Fixed in [ed299e5](https://github.com/MetaMask/metamask-extension/pull/20554/commits/ed299e5e080402f7b42a2f459c7a945ad93b0248)